File: android/src/main/java/io/curity/haapi/react/HaapiModule.kt
/*
 *  Copyright 2024 Curity AB
 *
 *  Licensed under the Apache License, Version 2.0 (the "License");
 *  you may not use this file except in compliance with the License.
 *  You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the License is distributed on an "AS IS" BASIS,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the License for the specific language governing permissions and
 *  limitations under the License.
 */
package io.curity.haapi.react

import android.util.Log
import com.facebook.react.bridge.LifecycleEventListener
import com.facebook.react.bridge.Promise
import com.facebook.react.bridge.ReactApplicationContext
import com.facebook.react.bridge.ReactContextBaseJavaModule
import com.facebook.react.bridge.ReactMethod
import com.facebook.react.bridge.ReadableMap
import com.google.gson.Gson
import io.curity.haapi.react.JsonUtil.jsonToNativeMap
import io.curity.haapi.react.JsonUtil.mapToNativeMap
import io.curity.haapi.react.events.EventEmitter
import io.curity.haapi.react.events.EventType
import io.curity.haapi.react.events.EventType.*
import io.curity.haapi.react.webauthn.WebAuthnHandler
import se.curity.identityserver.haapi.android.driver.HaapiLogger
import se.curity.identityserver.haapi.android.sdk.models.AuthenticatorSelectorStep
import se.curity.identityserver.haapi.android.sdk.models.ContinueSameStep
import se.curity.identityserver.haapi.android.sdk.models.HaapiRepresentation
import se.curity.identityserver.haapi.android.sdk.models.HaapiResponse
import se.curity.identityserver.haapi.android.sdk.models.Link
import se.curity.identityserver.haapi.android.sdk.models.OAuthAuthorizationResponseStep
import se.curity.identityserver.haapi.android.sdk.models.PollingStatus
import se.curity.identityserver.haapi.android.sdk.models.PollingStep
import se.curity.identityserver.haapi.android.sdk.models.ProblemRepresentation
import se.curity.identityserver.haapi.android.sdk.models.ProblemType
import se.curity.identityserver.haapi.android.sdk.models.RepresentationType
import se.curity.identityserver.haapi.android.sdk.models.WebAuthnAuthenticationClientOperationStep
import se.curity.identityserver.haapi.android.sdk.models.WebAuthnRegistrationClientOperationStep
import se.curity.identityserver.haapi.android.sdk.models.actions.Action
import se.curity.identityserver.haapi.android.sdk.models.actions.FormActionModel
import se.curity.identityserver.haapi.android.sdk.models.oauth.ErrorTokenResponse
import se.curity.identityserver.haapi.android.sdk.models.oauth.SuccessfulTokenResponse
import se.curity.identityserver.haapi.android.sdk.models.oauth.TokenResponse

const val TAG = "HaapiNative"

class HaapiModule(private val _reactContext: ReactApplicationContext) : ReactContextBaseJavaModule(_reactContext),
    LifecycleEventListener {

    override fun getName() = "HaapiModule"

    private var _haapiResponse: HaapiRepresentation? = null
    private var _tokenResponse: SuccessfulTokenResponse? = null
    private val _gson = Gson()
    private val _handler = HaapiInteractionHandler(_reactContext)
    private val _eventEmitter = EventEmitter(_reactContext)
    private val _webAuthnHandler = WebAuthnHandler(_reactContext)

    init {
        HaapiLogger.enabled = true
        HaapiLogger.isDebugEnabled = true
        _reactContext.addLifecycleEventListener(this)
    }

    override fun onHostResume() {
        Log.d(TAG, "App was resumed")
    }

    override fun onHostPause() {
        Log.d(TAG, "App was paused")
    }

    override fun onHostDestroy() {
        Log.d(TAG, "App was destroyed")
    }

    @ReactMethod
    fun load(conf: ReadableMap, promise: Promise) {
        try {
            _handler.loadConfiguration(conf.toHashMap())
        } catch (e: HaapiException) {
            rejectRequest(e, promise)
        }
        promise.resolve(true)
    }

    @ReactMethod
    fun start(promise: Promise) {
        Log.d(TAG, "Start was called")

        try {
            _handler.startAuthentication { response -> handleHaapiResponse(response, promise) }
        } catch (e: Exception) {
            Log.e(TAG, e.message ?: "Failed to attest $e")
            rejectRequest(e, promise)
        }
    }

    @ReactMethod
    fun logout(promise: Promise) {
        Log.d(TAG, "Logout was called, revoking tokens")

        if (_tokenResponse != null) {
            _handler.logoutAndRevokeTokens(_tokenResponse!!.accessToken, _tokenResponse!!.refreshToken)
        } else {
            _handler.closeHaapiConnection()
        }

        _tokenResponse = null
        resolveRequest(LoggedOut, "{}", promise)
    }

    @ReactMethod
    fun refreshAccessToken(refreshToken: String, promise: Promise) {
        Log.d(TAG, "Refreshing access token")

        try {
            _handler.refreshAccessToken(refreshToken) { tokenResponse ->
                handleTokenResponse(tokenResponse, promise)
            }
        } catch (e: Exception) {
            Log.d(TAG, "Failed to revoke tokens: ${e.message}")
            rejectRequest(e, promise)
        }
    }

    @ReactMethod
    fun navigate(linkMap: ReadableMap, promise: Promise) {

        val linkJson = _gson.toJson(linkMap.toHashMap())
        val link = _gson.fromJson(linkJson, Link::class.java)
        try {
            _handler.followLink(link) { response -> handleHaapiResponse(response, promise) }
        } catch (e: Exception) {
            Log.d(TAG, "Failed to navigate to link: ${e.message}")
            rejectRequest(e, promise)
        }
    }

    @ReactMethod
    fun submitForm(actionMap: ReadableMap, parameters: ReadableMap, promise: Promise) {

        val action = findAction(actionMap, _haapiResponse as HaapiRepresentation)
        if (action == null) {
            Log.d(TAG, "Failed to find action to submit. Possible re-submit")
            return
        }

        submitModel(action.model, parameters.toHashMap(), promise)
    }

    /**
     * Needs to be here to suppress warnings from react-native
     */
    @ReactMethod
    fun addListener(eventName: String) {

    }

    /**
     * Needs to be here to suppress warnings from react-native
     */
    @ReactMethod
    fun removeListeners(count: Int) {

    }

    private fun submitModel(model: FormActionModel, parameters: Map<String, Any>, promise: Promise) {
        Log.d(TAG, "Submitting form $model")
        try {
            _handler.submitForm(model, parameters) { response ->
                handleHaapiResponse(response, promise)
            }
        } catch (e: Exception) {
            Log.w(TAG, "Failed to submit form: ${e.message}")
            rejectRequest(e, promise)
        }

    }

    private fun handleCodeResponse(response: OAuthAuthorizationResponseStep, promise: Promise) {
        try {
            _handler.exchangeCode(response) { tokenResponse ->
                handleTokenResponse(tokenResponse, promise)
            }

        } catch (e: Exception) {
            Log.w(TAG, "Failed to exchange code: ${e.message}")
            rejectRequest(e, promise)
        }
    }

    private fun handleTokenResponse(tokenResponse: TokenResponse, promise: Promise) {

        if (tokenResponse is SuccessfulTokenResponse) {
            val tokenMap = mapOf(
                "accessToken" to tokenResponse.accessToken,
                "refreshToken" to tokenResponse.refreshToken,
                "scope" to tokenResponse.scope,
                "idToken" to tokenResponse.idToken,
                "expiresIn" to tokenResponse.expiresIn.seconds
            )
            _tokenResponse = tokenResponse
            resolveRequest(TokenResponse, _gson.toJson(tokenMap), promise)
        } else {
            tokenResponse as ErrorTokenResponse
            val errorResponseMap = mapOf(
                "error" to tokenResponse.error, "error_description" to tokenResponse.errorDescription
            )
            resolveRequest(TokenResponseError, JsonUtil.toJsonString(errorResponseMap), promise)
        }
    }

    private fun findAction(
        actionMap: ReadableMap,
        representation: HaapiRepresentation,
    ): Action.Form? {
        // TODO: Match more precise
        val submitKind = actionMap.getString("kind")
        Log.d(TAG, "Looking for form with kind $submitKind")
        Log.d(TAG, "Current response: $_haapiResponse")

        val action = representation.actions.find {
            it.toJsonObject().get("kind") == submitKind
        }
        return action as? Action.Form
    }

    private fun handleHaapiResponse(response: HaapiResponse, promise: Promise) {

        when (response) {
            is HaapiRepresentation -> {
                Log.d(TAG, "Updating reference to representation")
                _haapiResponse = response

                when (response) {
                    is WebAuthnRegistrationClientOperationStep -> handleWebAuthnRegistration(response, promise)

                    is WebAuthnAuthenticationClientOperationStep -> handleWebAuthnAuthentication(response, promise)

                    is AuthenticatorSelectorStep -> resolveRequest(
                        AuthenticationSelectorStep, response.toJsonString(), promise
                    )

                    is ContinueSameStep -> resolveRequest(
                        ContinueSameStep, response.toJsonString(), promise
                    )

                    is OAuthAuthorizationResponseStep -> handleCodeResponse(response, promise)
                    is PollingStep -> handlePollingStep(response, promise)

                    else -> when (response.type) {
                        RepresentationType.AUTHENTICATION_STEP -> {
                            resolveRequest(AuthenticationStep, response.toJsonString(), promise)
                        }

                        RepresentationType.REGISTRATION_STEP -> {
                            resolveRequest(RegistrationStep, response.toJsonString(), promise)
                        }

                        else -> {
                            Log.d(TAG, "Unknown step ${response.type}")
                            resolveRequest(UnknownResponse, response.toJsonString(), promise)
                        }
                    }
                }
            }

            is ProblemRepresentation -> {
                when (response.type) {
                    ProblemType.IncorrectCredentialsProblem -> {
                        resolveRequest(IncorrectCredentials, response.toJsonString(), promise)
                    }

                    ProblemType.SessionAndAccessTokenMismatchProblem -> {
                        resolveRequest(SessionTimedOut, response.toJsonString(), promise)
                    }

                    else -> {
                        resolveRequest(ProblemRepresentation, response.toJsonString(), promise)
                    }
                }
            }
        }

        Log.d(TAG, response.toJsonString())
    }

    private fun handleWebAuthnAuthentication(
        response: WebAuthnAuthenticationClientOperationStep,
        promise: Promise
    ) =
        _webAuthnHandler.authenticationFlow(response, promise) { parameters ->
            submitModel(response.continueFormActionModel, parameters, promise)
        }

    private fun handleWebAuthnRegistration(
        response: WebAuthnRegistrationClientOperationStep,
        promise: Promise
    ) =
        _webAuthnHandler.registrationFlow(response, promise) { parameters ->
            submitModel(response.continueFormActionModel, parameters, promise)
        }

    private fun handlePollingStep(pollingStep: PollingStep, promise: Promise) {

        when (pollingStep.properties.status) {
            PollingStatus.PENDING -> {
                _eventEmitter.sendEvent(PollingStep, pollingStep.toJsonString())
                resolveRequest(PollingStepResult, pollingStep.toJsonString(), promise)
            }

            PollingStatus.FAILED -> {
                _eventEmitter.sendEvent(StopPolling, "{}")
                submitModel(pollingStep.mainAction.model, emptyMap(), promise)
            }

            PollingStatus.DONE -> {
                _eventEmitter.sendEvent(StopPolling, "{}")
                submitModel(pollingStep.mainAction.model, emptyMap(), promise)
            }
        }
    }

    private fun rejectRequest(exception: Exception, promise: Promise) {
        val errorDescription = exception.message ?: "general error"
        val jsonMap = mutableMapOf("error" to "HaapiError", "error_description" to errorDescription)

        Log.d(TAG, "Rejecting request using with description '$errorDescription'")

        _eventEmitter.sendEvent(HaapiError, JsonUtil.toJsonString(jsonMap))
        promise.reject(exception, mapToNativeMap(jsonMap))
    }

    private fun resolveRequest(eventType: EventType, body: String, promise: Promise) {
        Log.d(TAG, "Resolving request using event type $eventType and body: $body")

        _eventEmitter.sendEvent(eventType, body)
        promise.resolve(jsonToNativeMap(body))
    }
}

File: ios/HaapiModule/HaapiModule.swift
//
// Copyright (C) 2024 Curity AB.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//   http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
//

import Foundation
import IdsvrHaapiSdk
import OSLog

@objc(HaapiModule)
class HaapiModule: RCTEventEmitter {

    struct Promise {
        var resolve: RCTPromiseResolveBlock
        var reject: RCTPromiseRejectBlock
    }

    enum InitializationError: Error {
        case moduleNotLoaded(msg: String = "Configuration not created. Run load() on the module before starting a flow")
        case failedToCreateHaapiManager(_ rootCause: Error)
        case failedToCreateOAuthTokenManager(_ rootCause: Error)
    }

    private var currentRepresentation: HaapiRepresentation?
    private var haapiConfiguration: HaapiConfiguration?

    private let logger = Logger()
    private let jsonDecoder: JSONDecoder = JSONDecoder()
    private let jsonEncoder: JSONEncoder = JSONEncoder()

    private var backingHaapiManager: HaapiManager?
    private var backingTokenManager: OAuthTokenManager?

    private var haapiManager: HaapiManager  {
        get throws {
            if (backingHaapiManager != nil) {
                return backingHaapiManager!
            }

            if(haapiConfiguration == nil) {
                throw InitializationError.moduleNotLoaded()
            }

            do {
                backingHaapiManager = try HaapiManager(haapiConfiguration: haapiConfiguration!)
                return backingHaapiManager!

            } catch {
                throw InitializationError.failedToCreateHaapiManager(error)
            }
        }
    }

    private var oauthTokenManager: OAuthTokenManager {
        get throws {
            if backingTokenManager != nil {
                return backingTokenManager!
            }

            guard haapiConfiguration != nil else { throw InitializationError.moduleNotLoaded() }

            backingTokenManager = OAuthTokenManager(oauthTokenConfiguration: haapiConfiguration!)
            return backingTokenManager!
        }
    }

    override init() {
        super.init()
        HaapiLogger.followUpTags = DriverFollowUpTag.allCases + SdkFollowUpTag.allCases
    }

    override func supportedEvents() -> [String]! {
        return EventType.allCases.map { $0.rawValue }
    }

    @objc(load:resolver:rejecter:)
    func load(configuration: Dictionary<String, Any>,
              resolver resolve: @escaping RCTPromiseResolveBlock,
              rejecter reject: @escaping RCTPromiseRejectBlock) {
        let promise = Promise(resolve: resolve, reject: reject)

        do {
            haapiConfiguration = try ConfigurationHelper.createHaapiConfiguration(data: configuration)
            promise.resolve(true)
        } catch {
            rejectRequestWithError(description: "Could not configure module: \(error)", promise: promise)
        }
    }


    @objc(start:rejecter:)
    func start(resolver resolve: @escaping RCTPromiseResolveBlock,
               rejecter reject: @escaping RCTPromiseRejectBlock) {

        let promise = Promise(resolve: resolve, reject: reject)

        if(haapiConfiguration == nil) {
            rejectRequestWithError(description: "Configuration not created. Run load() on the module before starting a flow", promise: promise)
            return
        }

        if(backingHaapiManager != nil) {
            closeManagers()
        }

        sendHaapiEvent(EventType.HaapiLoading, body: ["loading": true], promise: promise)

        do {
            try haapiManager.start(completionHandler: { haapiResult in self.processHaapiResult(haapiResult, promise: promise) })
        } catch {
            rejectRequestWithError(description: "Failed to start authentication flow. Error: \(error)", promise: promise)
            return
        }
    }

    @objc
    func submitForm(_ action: Dictionary<String, Any>,
                    parameters: Dictionary<String, Any>,
                    resolver resolve: @escaping RCTPromiseResolveBlock,
                    rejecter reject: @escaping RCTPromiseRejectBlock) {
        let promise = Promise(resolve: resolve, reject: reject)

        guard let model = action["model"] else {
            rejectRequestWithError(description: "", promise: promise)
            return
        }

        do {
            let actionObject = try JSONSerialization.data(withJSONObject: model)
            let formActionModel = try jsonDecoder.decode(FormActionModel.self, from: actionObject)
            sendHaapiEvent(EventType.HaapiLoading, body: ["loading": true], promise: promise)
            try haapiManager.submitForm(formActionModel, parameters: parameters, completionHandler: { haapiResult in self.processHaapiResult(haapiResult, promise: promise) })
        } catch {
            rejectRequestWithError(description: "Failed to construct form to submit: \(error)", promise: promise)
        }
    }

    @objc
    func navigate(_ linkMap: Dictionary<String, Any>,
                  resolver resolve: @escaping RCTPromiseResolveBlock,
                  rejecter reject: @escaping RCTPromiseRejectBlock) {

        var mutableLinkMap = linkMap
        // should be optional in SDK
        if(mutableLinkMap["rel"] == nil) {
            mutableLinkMap["rel"] = "link"
        }

        let promise = Promise(resolve: resolve, reject: reject)

        do {
            let linkObject = try JSONSerialization.data(withJSONObject: mutableLinkMap)
            let link = try jsonDecoder.decode(Link.self, from: linkObject)
            try haapiManager.followLink(link, completionHandler: { haapiResult in self.processHaapiResult(haapiResult, promise: promise) })
        } catch {
            rejectRequestWithError(description: "Failed to construct link: \(error)", promise: promise)
        }
    }

    @objc
    func refreshAccessToken(_ refreshToken: String,
                            resolver resolve: @escaping RCTPromiseResolveBlock,
                            rejecter reject: @escaping RCTPromiseRejectBlock) {
        let promise = Promise(resolve: resolve, reject: reject)
        sendHaapiEvent(EventType.HaapiLoading, body: ["loading": true], promise: promise)
        do {
            try oauthTokenManager.refreshAccessToken(with: refreshToken, completionHandler: { tokenResponse in
                self.handle(tokenResponse: tokenResponse, promise: promise)
            })
        } catch {
            rejectRequestWithError(description: "Could not refresh access token: \(error)", promise: promise)
        }
    }

    @objc(logout:rejecter:)
    func logout(resolver resolve: @escaping RCTPromiseResolveBlock,
                rejecter reject: @escaping RCTPromiseRejectBlock) {
        closeManagers()
        resolveRequest(eventType: EventType.LoggedOut, body: ["loggedout": true], promise: Promise(resolve: resolve, reject: reject))
    }

    override static func requiresMainQueueSetup() -> Bool {
        return true
    }

    private func closeManagers() {
        backingHaapiManager?.close()
        backingHaapiManager = nil
        backingTokenManager = nil
    }

    private func processHaapiResult(_ haapiResult: HaapiResult, promise: Promise) {
        switch haapiResult {
        case .representation(let representation):
            do {
                try process(haapiRepresentation: representation, promise: promise)
            } catch {
                rejectRequestWithError(description: "Failed to process HAAPI representation: \(error)", promise: promise)
            }
        case .problem(let problemRepresentation):
            process(problemRepresentation: problemRepresentation, promise: promise)
        case .error(let error):
            rejectRequestWithError(description: "Unknown error: " + error.localizedDescription, promise: promise)
        }
    }

    private func process(problemRepresentation: ProblemRepresentation, promise: Promise) {
        switch(problemRepresentation.type) {
        case .incorrectCredentialsProblem:
            resolveRequest(eventType: EventType.IncorrectCredentials, body: problemRepresentation, promise: promise)
        case .sessionAndAccessTokenMismatchProblem:
            resolveRequest(eventType: EventType.SessionTimedOut, body: problemRepresentation, promise: promise)
        default:
            resolveRequest(eventType: EventType.ProblemRepresentation, body: problemRepresentation, promise: promise)
        }
    }

    private func process(haapiRepresentation: HaapiRepresentation, promise: Promise) throws {
        switch(haapiRepresentation) {
        case let step as WebAuthnAuthenticationClientOperationStep:
            handle(webauthnStep: step, promise: promise)
        case let step as WebAuthnRegistrationClientOperationStep:
            handle(webauthnRegistrationStep: step,  promise: promise)
        case is AuthenticatorSelectorStep:
            resolveRequest(eventType: EventType.AuthenticationSelectorStep, body: haapiRepresentation, promise: promise)
        case is InteractiveFormStep:
            resolveRequest(eventType: EventType.AuthenticationStep, body: haapiRepresentation, promise: promise)
        case is ContinueSameStep:
            resolveRequest(eventType: EventType.ContinueSameStep, body: haapiRepresentation, promise: promise)
        case let step as OAuthAuthorizationResponseStep:
            try handle(codeStep: step, promise: promise)
        case let step as PollingStep:
            try handle(pollingStep: step, promise: promise)
        default:
            rejectRequestWithError(description: "Unknown step", promise: promise)
        }
    }

    private func handle(webauthnRegistrationStep: WebAuthnRegistrationClientOperationStep,
                        promise: Promise) {
        logger.debug("Handle webauthn registration step")
        // Start passkey registration
        if #available(iOS 15.0, *) {
            WebAuthnClientOperationHandler().register(operation: webauthnRegistrationStep) { result in
                switch result {
                case .success(let credential):
                    guard let credentialOptions = webauthnRegistrationStep.actionModel.platformOptions else {
                        return self.rejectRequestWithError(description: "Failed to got credential options", promise:promise)
                    }

                    let attestationObject = credential.rawAttestationObject ?? Data()

                    let parameters = webauthnRegistrationStep.formattedParametersForRegistration(credentialOptions: credentialOptions,
                                                                                                 attestationObject: attestationObject,
                                                                                                 rawClientDataJSON: credential.rawClientDataJSON,
                                                                                                 credentialID: credential.credentialID)
                    self.submitModel(model: webauthnRegistrationStep.continueAction.model, parameters: parameters, promise: promise)
                case .failure(.userCancelled):
                    self.logger.debug("User cancelled the webauthn registration dialog")
                    self.resolveRequest(eventType: EventType.WebAuthnUserCancelled, body: webauthnRegistrationStep, promise: promise)
                case .failure(let error):
                    self.logger.info("WebAuthn registration failed: \(error)")
                    self.resolveRequest(eventType: EventType.WebAuthnRegistrationFailed, body: webauthnRegistrationStep, promise: promise)
                }
            }
        } else {
            rejectRequestWithError(description: "Passkeys are not supported on OS version before 15.0", promise: promise)
        }
    }

    private func handle(webauthnStep: WebAuthnAuthenticationClientOperationStep,
                        promise: Promise) {
        if #available(iOS 15.0, *) {
            WebAuthnClientOperationHandler().authenticate(operation: webauthnStep) { result in
                switch result {
                case .success(let assertion):
                    let parameters = webauthnStep.formattedParametersForAssertion(rawAuthenticatorData: assertion.rawAuthenticatorData,
                                                                                  rawClientDataJSON: assertion.rawClientDataJSON,
                                                                                  signature: assertion.signature,
                                                                                  credentialID: assertion.credentialID)
                    self.submitModel(model: webauthnStep.continueAction.model, parameters: parameters, promise: promise)
                case .failure(.userCancelled):
                    self.logger.debug("User cancelled the authentication dialog")
                    self.resolveRequest(eventType: EventType.WebAuthnUserCancelled, body: webauthnStep, promise: promise)
                case .failure(let error):
                    self.logger.info("WebAuthn authentication failed: \(error)")
                    self.resolveRequest(eventType: EventType.WebAuthnAuthenticationFailed, body: webauthnStep, promise: promise)
                }
            }
        } else {
            rejectRequestWithError(description: "Passkeys are not supported on OS version before 15.0", promise: promise)
        }
    }

    private func handle(pollingStep: PollingStep,
                        promise: Promise) throws {
        sendHaapiEvent(EventType.PollingStep, body: pollingStep, promise: promise)

        switch(pollingStep.pollingProperties.status) {
        case .pending:
            resolveRequest(eventType: EventType.PollingStepResult, body: pollingStep, promise: promise)
        case .failed:
            sendHaapiEvent(EventType.StopPolling, body: pollingStep, promise: promise)
            submitModel(model: pollingStep.mainAction.model, promise: promise)
        case .done:
            sendHaapiEvent(EventType.StopPolling, body: pollingStep, promise: promise)
            submitModel(model: pollingStep.mainAction.model, promise: promise)
        }
    }

    private func handle(tokenResponse: TokenResponse, promise: Promise) {
        switch(tokenResponse) {
        case .successfulToken(let successfulTokenResponse):
            let tokenResponse = SuccessTokenResponse(successfulTokenResponse)
            resolveRequest(eventType: EventType.TokenResponse, body: tokenResponse, promise: promise)
        case .errorToken(let errorTokenResponse):
            // Request succeeded, but with contents indicating an error. Resolve with contents, so that frontend can act on it.
            resolveRequest(eventType: EventType.TokenResponseError, body: errorTokenResponse, promise: promise)
        case .error:
            rejectRequestWithError(description: "Failed to execute token request", promise: promise)
        }
    }

    private func submitModel(model: FormActionModel,
                             parameters: [String: Any] = [:],
                             promise: Promise) {
        do {
            try haapiManager.submitForm(model, parameters: parameters, completionHandler: { haapiResult in
                self.processHaapiResult(haapiResult, promise: promise)
            })
        }
        catch {
            rejectRequestWithError(description: "Failed to submit model", promise: promise)
        }
    }

    private func handle(codeStep: OAuthAuthorizationResponseStep, promise: Promise) throws {
        try oauthTokenManager.fetchAccessToken(with: codeStep.oauthAuthorizationResponseProperties.code!, dpop: haapiManager.dpop, completionHandler: { tokenResponse in
            self.handle(tokenResponse: tokenResponse, promise: promise)
        })
    }

    private func sendHaapiEvent(_ type: EventType, body: Codable, promise: Promise) {
        do {
            let encodedBody = try encodeObject(body)
            logger.debug("Sending event: \(type.rawValue)")
            self.sendEvent(withName: type.rawValue, body: encodedBody)
        }
        catch {
            rejectRequestWithError(description: "Could not encode event as json. Error: \(error)", promise: promise)
        }
    }

    private func encodeObject(_ object: Codable) throws -> Any {
        do {
            let jsonData = try jsonEncoder.encode(object)
            return try JSONSerialization.jsonObject(with: jsonData)
        }
        catch {
            throw NSError()
        }
    }

    private func rejectRequestWithError(description: String, promise: Promise) {
        sendHaapiEvent(EventType.HaapiError, body: ["error": "HaapiError", "error_description": description], promise: promise)
        sendHaapiEvent(EventType.HaapiFinishedLoading, body: ["loading": false], promise: promise)
        promise.reject("HaapiError", description, nil)
        closeManagers()
    }

    private func resolveRequest(eventType: EventType, body: Codable, promise: Promise) {
        sendHaapiEvent(EventType.HaapiFinishedLoading, body: ["loading": false], promise: promise)
        do {
            let encodedBody = try encodeObject(body)

            promise.resolve(encodedBody)
            self.sendHaapiEvent(eventType, body: body, promise: promise)
        }
        catch {
            rejectRequestWithError(description: "Could not encode response as json. Error: \(error)", promise: promise)
        }
    }

    private struct SuccessTokenResponse : Codable {
        var accessToken: String
        var refreshToken: String?
        var idToken: String?
        var scope: String
        var expiresIn: Int

        init(_ tokenResponse : SuccessfulTokenResponse) {
            self.accessToken = tokenResponse.accessToken
            self.idToken = tokenResponse.idToken
            self.refreshToken = tokenResponse.refreshToken
            self.scope = tokenResponse.scope ?? ""
            self.expiresIn = tokenResponse.expiresIn
        }
    }
}
